File: packages/typescript/src/surfaces/system.ts
import type { YaiEnvelope } from '../envelope';
import { YAI_OPERATIONS } from '../operations';
import { TransportBackedSurface } from './base';

export type RuntimeLifecycleState =
  | 'unknown'
  | 'unavailable'
  | 'stopped'
  | 'starting'
  | 'running'
  | 'stopping'
  | 'degraded'
  | 'error';

export type RuntimeServiceReadiness = 'ready' | 'pending' | 'unavailable' | 'degraded' | 'error';
export type RuntimeServiceMode = 'local-dev' | 'local-service-planned' | 'remote-service-planned';
export type RuntimeServiceControlPosture = 'dev-wrapper' | 'service-manager-planned' | 'unavailable';
export type RuntimeServiceManagerPosture = 'service-manager-planned' | 'unavailable';
export type RuntimeServiceInstallPosture = 'not-installed' | 'installed' | 'unavailable';
export type RuntimeOperationalReadiness = 'ready' | 'sealed' | 'pending' | 'blocked' | 'unavailable' | 'error';
export type RuntimeSealedPosture = 'sealed' | 'unsealed' | 'pending';
export type RuntimeAuthPosture = 'authenticated' | 'unauthenticated' | 'local-dev' | 'unavailable' | 'unknown';
export type RuntimeCasePosture =
  | 'root_present'
  | 'missing_root'
  | 'tree_present'
  | 'missing_tree'
  | 'root_present/tree_present'
  | 'unavailable'
  | 'unknown';
export type RuntimeOperatorContextPosture =
  | 'active_case_present'
  | 'active_case_missing'
  | 'unavailable'
  | 'pending'
  | 'planned'
  | 'available'
  | 'unknown';
export type RuntimeClientPosture = 'cli-one-shot' | 'loom-tui' | 'sdk-embedded' | 'unknown';
export type RuntimeSessionPosture = 'legacy_ignored' | 'legacy_compatibility' | 'unknown';
export type RuntimeControlPlanServiceManager = 'unavailable' | 'launchd' | 'systemd' | 'manual' | 'unknown';
export type RuntimeControlActionAvailability = 'available' | 'unavailable' | 'planned' | 'unsupported';
export type RuntimeControlActionReason =
  | 'service_manager_unavailable'
  | 'dev_wrapper_only'
  | 'not_implemented'
  | 'unsupported_platform'
  | 'insufficient_permission'
  | 'unknown';
export type RuntimeSealReason =
  | 'none'
  | 'missing_auth_context'
  | 'missing_root_case'
  | 'missing_case_tree'
  | 'missing_operator_context'
  | 'runtime_transport_unavailable'
  | 'unknown'
  | 'client_connection_unavailable'
  | 'operator_context_unavailable'
  | 'authorization_context_unavailable'
  | 'active_case_unselected'
  | 'unsealed';
export type RuntimeActiveCaseReason =
  | 'active_case_unselected'
  | 'active_case_unavailable'
  | 'active_case_blocked'
  | 'active_case_error'
  | 'none';

export interface RuntimeControlPlanAction {
  available: boolean;
  availability?: RuntimeControlActionAvailability;
  reason: RuntimeControlActionReason;
  executionClaim: boolean;
}

export interface RuntimeControlPlan {
  serviceManager: RuntimeControlPlanServiceManager;
  start: RuntimeControlPlanAction;
  stop: RuntimeControlPlanAction;
  restart: RuntimeControlPlanAction;
}

export interface RuntimeServiceStatus {
  schema: 'api.runtime.service.status.v1';
  serviceId?: string;
  status: RuntimeLifecycleState;
  lifecycle?: RuntimeLifecycleState;
  lifecycleState: RuntimeLifecycleState;
  readiness: RuntimeServiceReadiness;
  health: 'available' | 'pending' | 'unavailable' | 'degraded' | 'error' | 'unknown';
  transport?: string;
  sealed: boolean;
  sealedPosture: RuntimeSealedPosture;
  sealReason: RuntimeSealReason;
  authPosture?: RuntimeAuthPosture;
  casePosture?: RuntimeCasePosture;
  clientPosture?: RuntimeClientPosture;
  sessionPosture?: RuntimeSessionPosture;
  allowedSurfaces: Array<'health' | 'status' | 'readiness'>;
  blockedSurfaces: Array<'operational-actions'>;
  identityPosture: 'unavailable' | 'local-dev' | 'available';
  operatorContextPosture: RuntimeOperatorContextPosture;
  activeCasePosture: 'unavailable' | 'unselected' | 'pending' | 'planned' | 'selected';
  activeCaseReason: RuntimeActiveCaseReason;
  activeCaseRef: string;
  serviceLifecycle: RuntimeLifecycleState;
  serviceManagerPosture: RuntimeServiceManagerPosture;
  serviceControlPosture: RuntimeServiceControlPosture;
  serviceInstallPosture: RuntimeServiceInstallPosture;
  devWrapperStatus: RuntimeLifecycleState | 'pending' | 'unavailable';
  systemServiceStatus: 'running' | 'pending' | 'unavailable' | 'not-installed' | 'error';
  operationalReadiness: RuntimeOperationalReadiness;
  operationalReadinessReason: RuntimeSealReason | 'none';
  controlPlan?: RuntimeControlPlan;
  mode: RuntimeServiceMode;
  controlPosture: RuntimeServiceControlPosture;
  message: string;
  warnings?: string[];
  refs?: Record<string, string>;
}

function systemUnavailable(operationId: typeof YAI_OPERATIONS[keyof typeof YAI_OPERATIONS]): YaiEnvelope<RuntimeServiceStatus> {
  return {
    operation_id: operationId,
    status: 'unavailable',
    execution_claim: false,
    implementation_status: 'transport-unconfigured',
    message: 'System transport not configured; runtime/system status is unavailable until a real transport is attached'
  };
}

export class SystemSurface extends TransportBackedSurface {
  status(request?: unknown): Promise<YaiEnvelope<RuntimeServiceStatus>> {
    if (!this.transport) {
      return Promise.resolve(systemUnavailable(YAI_OPERATIONS.systemStatus));
    }
    return this.transport.invoke<RuntimeServiceStatus>(YAI_OPERATIONS.systemStatus, request);
  }

  check(request?: unknown): Promise<YaiEnvelope<RuntimeServiceStatus>> {
    if (!this.transport) {
      return Promise.resolve(systemUnavailable(YAI_OPERATIONS.systemCheck));
    }
    return this.transport.invoke<RuntimeServiceStatus>(YAI_OPERATIONS.systemCheck, request);
  }

  runtimeInspect(request?: unknown): Promise<YaiEnvelope<RuntimeServiceStatus>> {
    if (!this.transport) {
      return Promise.resolve(systemUnavailable(YAI_OPERATIONS.systemRuntimeInspect));
    }
    return this.transport.invoke<RuntimeServiceStatus>(YAI_OPERATIONS.systemRuntimeInspect, request);
  }
}

File: packages/rust/src/surfaces/system.rs
use serde::{Deserialize, Serialize};

use crate::envelope::YaiEnvelope;
use crate::operations::YAI_OPERATIONS;
use crate::status::YaiResult;
use crate::transport::YaiTransport;

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct RuntimeControlPlanAction {
    #[serde(default)]
    pub available: bool,
    #[serde(default)]
    pub availability: Option<String>,
    #[serde(default)]
    pub reason: Option<String>,
    #[serde(rename = "executionClaim", alias = "execution_claim", default)]
    pub execution_claim: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct RuntimeControlPlan {
    #[serde(rename = "serviceManager", alias = "service_manager", default)]
    pub service_manager: Option<String>,
    #[serde(default)]
    pub start: Option<RuntimeControlPlanAction>,
    #[serde(default)]
    pub stop: Option<RuntimeControlPlanAction>,
    #[serde(default)]
    pub restart: Option<RuntimeControlPlanAction>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RuntimeServiceStatus {
    #[serde(default)]
    pub schema: Option<String>,
    pub status: String,
    #[serde(rename = "lifecycleState", alias = "lifecycle_state", default)]
    pub lifecycle_state: Option<String>,
    #[serde(default)]
    pub lifecycle: Option<String>,
    #[serde(default)]
    pub readiness: Option<String>,
    #[serde(default)]
    pub health: Option<String>,
    #[serde(default)]
    pub transport: Option<String>,
    #[serde(
        rename = "operationalReadiness",
        alias = "operational_readiness",
        default
    )]
    pub operational_readiness: Option<String>,
    #[serde(rename = "sealReason", alias = "seal_reason", default)]
    pub seal_reason: Option<String>,
    #[serde(rename = "authPosture", alias = "auth_posture", default)]
    pub auth_posture: Option<String>,
    #[serde(rename = "casePosture", alias = "case_posture", default)]
    pub case_posture: Option<String>,
    #[serde(
        rename = "operatorContextPosture",
        alias = "operator_context_posture",
        default
    )]
    pub operator_context_posture: Option<String>,
    #[serde(rename = "clientPosture", alias = "client_posture", default)]
    pub client_posture: Option<String>,
    #[serde(rename = "sessionPosture", alias = "session_posture", default)]
    pub session_posture: Option<String>,
    #[serde(rename = "controlPlan", alias = "control_plan", default)]
    pub control_plan: Option<RuntimeControlPlan>,
    #[serde(default)]
    pub message: Option<String>,
}

pub struct SystemSurface<TTransport: YaiTransport> {
    transport: TTransport,
}

impl<TTransport: YaiTransport> SystemSurface<TTransport> {
    pub fn new(transport: TTransport) -> Self {
        Self { transport }
    }

    pub fn status(&self) -> YaiResult<YaiEnvelope<RuntimeServiceStatus>> {
        self.transport
            .invoke::<RuntimeServiceStatus>(YAI_OPERATIONS.system_status, None)
    }

    pub fn check(&self) -> YaiResult<YaiEnvelope<RuntimeServiceStatus>> {
        self.transport
            .invoke::<RuntimeServiceStatus>(YAI_OPERATIONS.system_check, None)
    }

    pub fn runtime_inspect(&self) -> YaiResult<YaiEnvelope<RuntimeServiceStatus>> {
        self.transport
            .invoke::<RuntimeServiceStatus>(YAI_OPERATIONS.system_runtime_inspect, None)
    }
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use crate::{NotConfiguredTransport, YaiClient, YaiError};

    use super::RuntimeServiceStatus;

    #[test]
    fn system_status_not_configured_is_not_success() {
        let client = YaiClient::new(NotConfiguredTransport);
        let result = client.system().status();
        assert!(matches!(
            result,
            Err(YaiError::TransportNotConfigured {
                operation_id: "system.status"
            })
        ));
    }

    #[test]
    fn system_status_deserializes_canonical_runtime_posture_fields() {
        let data: RuntimeServiceStatus = serde_json::from_value(json!({
            "schema": "api.runtime.service.status.v1",
            "status": "unavailable",
            "lifecycle": "unavailable",
            "lifecycleState": "unavailable",
            "health": "unavailable",
            "readiness": "unavailable",
            "transport": "runtime_transport_unavailable",
            "operationalReadiness": "sealed",
            "sealReason": "missing_auth_context",
            "authPosture": "unauthenticated",
            "casePosture": "unavailable",
            "operatorContextPosture": "unavailable",
            "clientPosture": "sdk-embedded",
            "sessionPosture": "legacy_ignored",
            "controlPlan": {
                "serviceManager": "unavailable",
                "start": {
                    "available": false,
                    "availability": "unavailable",
                    "reason": "not_implemented",
                    "executionClaim": false
                },
                "stop": {
                    "available": false,
                    "availability": "unavailable",
                    "reason": "not_implemented",
                    "executionClaim": false
                },
                "restart": {
                    "available": false,
                    "availability": "unavailable",
                    "reason": "not_implemented",
                    "executionClaim": false
                }
            }
        }))
        .expect("canonical runtime posture fields should deserialize");

        assert_eq!(data.operational_readiness.as_deref(), Some("sealed"));
        assert_eq!(data.seal_reason.as_deref(), Some("missing_auth_context"));
        assert_eq!(data.client_posture.as_deref(), Some("sdk-embedded"));
        assert_eq!(
            data.control_plan
                .as_ref()
                .and_then(|plan| plan.start.as_ref())
                .map(|action| action.execution_claim),
            Some(false)
        );
    }
}
